### ISSUE TITLE
Telemetry / (anonymous) statistics about number of app installs in the wild?

### ISSUE THREAD

**Issue by alexAubin** [YunoHost/issues#1670]
Being able to obtain quantitative data about the number of times an app is installed on servers is a recurring topic. The discussion usually ends with : 
- we could do it maybe, but it would be opt-in because privacy concerns etc
- we could do it but who has time for it ?

I'm personally not too enthusiast about an opt-in only mechanism because the data we get would be ultimately partial and we would have to extrapolate and speculate about the data, missing corners etc... (though I do agree with the privacy concerns and various implications)

Rediscussing this today with ljf and I was bringing the fact that "if we had a way to obtain the information without disclosing any private info like IP address / domain name of the actual server, we could legitimately design this as opt-out ?"

We ended up thinking about a design where an info is able to push some info to the yunohost infrastructure *without disclosing its address IP*, using DNS queries and DNS resolvers as "anonymizing proxies". Here's a draft of how it could work : 

- Regularly (say once every week), every yunohost instance sends some data
- To send the data, the instance will issue a serie of DNS request to regular DNS resolvers
- The queries are following the template `<some_payload>.v1.stats.yunohost.org` (here, `stats.yunohost.org` is an arbitrary choice that can be changed)
- Since the DNS resolver cannot resolve this, it asks the authority for stats.yunohost.org to handle the resolution
- The authority corresponds to a DNS server on our infrastructure that therefores receive the payload and is able to log it
- The payload : 
    - has to be relatively small because DNS queries are limited to ~250ish bytes - hence the need to probably have several of them to send all the data
    - shall be encrypted using asymetric keys to prevent the intermediate DNS resolvers from snooping on the data
    - contains some info related to one or several apps installed

From this, we should be able to compute how many times an app is installed in the wild without being able to tell which server exactly is running it

Advanced questions can include : 
- wether or not to report every apps independently from each other : depends on wether or not we want to be able to compute some correlation (app X installed with app Y as well), or if we think that may allow us to create some sort of "fingerprinting" of servers because we'd know the list of all apps installed which may be unique
- we may also want to include some sort of server UUID to be able to fight possible abuse (depending on wether or not we think it's a concern)

**Comment by zamentur** [comment]
Here a list of data to transfer through this mechanism:

- list of app_id
- nb of ynh users ?
- arch ?
- ram total / ram usage (rounded) ?
- disk total / disk usage (rounded) ?

**Comment by yalh76** [comment]
My opinion, do as Debian do it: 
- having a question during postinstall (like done about SSH changes), asking if user wants to participate to the statistics Yes/No
- this question will also validate the debian popularity-contest (proposition)
- for automatic install (LXC) response would be No
- adding a link in the https;//mydomain/yunohost/admin/#/update to let admin enable or disable statistics
- also having the question in the webadmin at https://mydomain/yunohost/admin/#/ until the user has approved or rejected the statistics

**Comment by Limezy** [comment]
Hi Alex and Ynh team,
Your proposal is very elegant and we need to find a way to explain to the Yunohost admin how committed you are towards anonymizing data and making sure the statistics don't jeopardize anything from the privacy, which is probably one of the main motivations for self-hosting !

**Comment by Salamandar** [comment]
I wrote my ideas here just today, not knowing there already was an issue about that : https://github.com/YunoHost/issues/discussions/1921

**Comment by Salamandar** [comment]
I'm not sure that hiding the IP address is a real issue.
* If i understand the code correctly, each yunohost already "phones home" to download the list of apps :
  https://github.com/YunoHost/yunohost/blob/dev/src/yunohost/app_catalog.py#L23
  We could use this domain to derive the one used to push statistics, to "protect" instances that changed this domain
* The same goes for the diagnosis tool. It pushes JSon data to diagnosis.yunohost.org every day.
* If the server is correctly configured, the ip addresses can be not-logged.

Don't mistake me here : I'm not trying to throw rocks at anyone about this behaviour. I'm trusting the people accessing the yunohost servers to handle the logs and ip addresses with care.

**Comment by alexAubin** [comment]
Yeah it's indeed perfectly possible that we're overthinking and over-engineering here :P

The proposed trick using DNS query could be a nice clever hack, but also nightmare to implement because of the size of the payload being quite small (and willingness to encrypt the data to not get snooped by the intermediate DNS resolver). (But maybe somewhere, somebody already solved the problem of anonymous metrics idk)

Anyway the main goal was basically to provide a technical guarantee that we can't build a map of what server actually hosts what app, how many users etc ... There's a big difference between just serving the app json (or even the diagnosis server) and obtaining per-IP the list of apps installed and other information you mention in your other thread... I mean even if we create such a service without logging the IP, it would be a matter of adding 2 lines of code on our server to also log it, and obtain a full map of who hosts what, which is a big deal. Note that it's not just about trusting us with this, but it's also about trusting that we won't get hacked by an attacker who would maybe be interested in such information, or be forced by legal authorities to deliver such informations ...

Under such configuration, personally I would never vote for such a system to be opt-out (or opt-in-but-encourage-people-to-opt-in) whereas an anonymous-by-design system would be different. (Though you could also argue that depending on the amount of data received, it's possible to partially unanomyze them idk)

aaaaaa overthinking again

In the meantime a not-perfect opt-in system would still be better than nothing at all ¯\\\_(ツ)\_/¯

**Comment by Salamandar** [comment]
> to provide a technical guarantee that we can't build a map of what server actually hosts what app, how many users etc

Yeah, that's the main issue. The way Home Assistant does it makes Cloudflare a "man in the middle" to prevent the Hass admins to know the ip addresses.

> In the meantime a not-perfect opt-in system would still be better than nothing at all ¯\_(ツ)_/¯

I 200% agree to this. In the end, the way data is transmitted is a small part of this project. A lot has to be implemented "downstream" and "upstream".

**Comment by alexAubin** [comment]
Yeah indeed, we should decouple the "anonymize transport" from the actual "implement API server/client" logic

Not a big fan of relying on Cloudflare, but I randomly searched for a lib that would allow to get a random HTTP proxy and lol turns out there's a lib for this ? :O https://github.com/jaredlgillespie/proxyscrape or https://github.com/jundymek/free-proxy (idk if that's what we want to use, but it sounds "doable" to anonymize traffic somehow even with an HTTP API)

**Comment by Salamandar** [comment]
What you're saying here is that a single server would have a HTTP proxy redirecting+anonymizing the requests on the fly ?
That's still better than nothing I guess.

Anyways, putting this "data transfer" issue aside, would you be ready to talk about implementing the rest of this "project" ?

**Comment by alexAubin** [comment]
I'm saying that on the client side, we could randomly pick a proxy HTTP server to relay the data and therefore anonymize it (by that I mean "we (the server) would not know the original IP")

Yes of course :+1:

**Comment by Salamandar** [comment]
Nice ! Then I may work on that.I'm thinking a daemon completely separated from Yunohost that calls the API, if that makes sense to you ?

**Comment by alexAubin** [comment]
yes :+1:

**Comment by Salamandar** [comment]
I started a biased implementation on https://github.com/Salamandar/yanalytics.
